From: Egbert Eich <eich@suse.de>
Subject: drm/edid: Fix EDID block that is shifted by up to 3 bytes.
Patch-mainline: never
References: bnc#712899 et.al.

This works around an issue seen in bnc#712899 where the EDID
block is shifted by one byte. This issue happens on some Radeon
ES1000 cards.
The underlying issue is not yet well understood, to accomodate
users we introduce this temporary workaround.
Once the bug is fixed it will be removed.

    Signed-off-by: Egbert Eich <eich@suse.de>
---
 drivers/gpu/drm/drm_edid.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- linux-3.0-SLE11-SP2-myfixes.orig/drivers/gpu/drm/drm_edid.c
+++ linux-3.0-SLE11-SP2-myfixes/drivers/gpu/drm/drm_edid.c
@@ -155,6 +155,16 @@ drm_edid_block_valid(u8 *raw_edid)
 	u8 csum = 0;
 	struct edid *edid = (struct edid *)raw_edid;
 
+	if (raw_edid[0] == 0xff) {
+	  int cnt = 1;
+	  while (raw_edid[cnt] == 0xff)
+	    cnt++;
+	  if (raw_edid[cnt] == 0x00 && cnt >= 3) {
+	    for (i = EDID_LENGTH; i >= (7 - cnt); i--)
+	      raw_edid[i] = raw_edid[i - 7 + cnt];
+	    memcpy(raw_edid, edid_header, sizeof(edid_header));
+	  }
+	}
 	if (raw_edid[0] == 0x00) {
 		int score = drm_edid_header_is_valid(raw_edid);
 		if (score == 8) ;
